### ISSUE TITLE
Passing a method's param value (read at run-time) in an InvokeExpr created in BodyTransformer?

### ISSUE THREAD

**Issue by amadionix** [soot-oss/soot#1551]
I have a class ClassWithMethods with a bunch of static methods and another class implementing a BodyTransformer called MyClass.

Inside MyClass, I create Stmt variables which I insert in the Unit chain. 
I create these Stmt variables by using this sintax:
`InvokeExpr expr = Jimple.v().newStaticInvokeExpr(method.makeRef());
 Stmt stmt = Jimple.v().newInvokeStmt(expr);`
where `method` is a SootMethod created out of a ClassWithMethods's method.

I am instrumenting a method and I would like to pass one of the input parameters of this instrumented method into the above `newStaticInvokeExpr`. How do I do this?

On a high level, what I am trying to do is to instrument a method to read its input parameters' values and store them in an in-memory data structure that I have in a static class somewhere.

**Comment by patricklam** [comment]
Hi Nicola,

At the start of the method you are instrumenting (assuming you have a
JimpleBody for it) you will find a number of IdentityStmts which assign
from parameters into normal Local variables. You can inspect these
IdentityStmts and see which local variable your parameter of interest ends
up in. Then you can put that in your expr that you are creating. Let me
know if that helps.

pat

**Comment by amadionix** [comment]
I see what you mean. It makes sense. It doesn't look like the easiest thing to do but I will try it out. Thank you!

**Comment by amadionix** [comment]
I have done what you suggested, which I think was finding the IdentityStms and then doing this:
`Value val = ((IdentityStmt) stmt).getLeftOp();` to get the actual value ref. 
First of all, is this right?

I still have an issue related to soot-class-path

I am doing this instrumentation in a method that implements a BodyTransformer. And every functionality of Soot works well inside there.
I have this external class with some static methods that I am putting in those expr.
At this point I have a method called `addFeature(Value val)`, that I would like to call with the value found above.

I get a soot-class-path error (`couldn't find class: soot.Value (is your soot-class-path set properly?`) at `freudLoggerClass = Scene.v().loadClassAndSupport("instrumenter.FreudLogger");`
where `instrumenter.FreudLogger` is the class containing `addFeature(Value val)`. 
Do you know why that could be?

This is how I am starting soot in the first place (maybe the problem is here):
`Pack jtp = PackManager.v().getPack("jtp");`
`jtp.add(new Transform("jtp.instrumenter", new InvokeStaticInstrumenter(args[1])));`
`soot.Main.main(new String[]{args[0]});`
`args[0]` and `args[1]` are two strings I'm passing to the instrumenter (they represent the names of the method that I want to instrument and that of the class that contains it).

[This](https://gitlab.software.imdea.org/nicola.amadio/freud-java-instrumentation) is the repository containing the classes I am working on. If you want to have a look there to get context, the relevant classes are inside the package `instrumenter` in `src/main/java`.

I have been working on this for days, I would really appreciate if you could please help me out!

**Comment by patricklam** [comment]
IdentityStmt: Yes, that sounds right.

OK, so it looks to me like you are now mixing compile-time context with
run-time context.

Soot.Value is a compile-time container for something that doesn't
necessarily exist at run-time. Assuming that val is a local variable of
type Local, you can create a staticInvokeExpr and pass it val. That local
variable has some declared type and it's that type you need to use as a
parameter for addFeature.

I feel like I could be more clear but I also would like to send this email
as quickly as possible... let me know if that makes sense or whether I
should try harder.

**Comment by amadionix** [comment]
OK on not passing the Value.

I don't understand why I should pass the type.
Let's say the type is `int`, what do I do with a String "int" inside my `addFeature`?

By the way, earlier I was trying to pass a `Soot.Local` to `addFeature` but that was giving me the soot-class-path error as well. (in case this is what you are suggesting).

Lastly, I feel like you said that I need to pass something different to `staticInvokeExpr` and `addFeature`? 
My understanding was that whatever I pass to the first should end up being a parameter to the second.

**Comment by patricklam** [comment]
addFeature should have a parameter of int type, not of Local type. You pass
the Local as a parameter to addFeature (by adding it to the
staticInvokeExpr) and Soot generates the actual parameter with the local.
Inside addFeature you have the int with the value you want.

**Comment by amadionix** [comment]
It works! Never thought it would work this way honestly... Kind of feels like magic.

Thank you so much for the help, you really made a difference!!!

**Comment by amadionix** [comment]
Is there a way to have `addFeature(String varName, int varValue)`?

If I have `addFeature(int varValue)` and I build `InvokeExpr expr = Jimple.v().newStaticInvokeExpr(addFeature.makeRef(), myLocal)`, it works. 

But I would also like to pass the variables' names.

**Comment by patricklam** [comment]
Glad to hear that helped. You can create a StringConstant and pass that
along to the function.

**Comment by amadionix** [comment]
`InvokeExpr expr = Jimple.v().newStaticInvokeExpr(addFeatureVal.makeRef(), Arrays.asList(new StringConstant(local.getName()), local));`

I've tried a bunch of different way to write this but I didn't manage to make it work. Any tip?

**Comment by patricklam** [comment]
Have you tried StringConstant.v(local.getName)?

**Comment by amadionix** [comment]
Right. Fixed that.

But know I get this other error:
`No method void addFeature(String, int) in class instrumenter.FreudLogger` in:
`addFeature = freudLoggerClass.getMethod("void addFeature(String, int)");`

I tried adding `java.lang` before String in the signature but it doesn't seem to make a difference.

**Comment by patricklam** [comment]
Not sure. You could try getMethodByName instead, or print out the list of
methods and see if what you expect to be there is there.

**Comment by amadionix** [comment]
Fixed also that. 

Problem now is that inside my addFeature(String name, int val), both name and val contain the actual value of that variable, while I would expect name to contain its name (local.getName()).

**Comment by patricklam** [comment]
Try pushing your code to your repo again... Hard to tell from here.

**Comment by amadionix** [comment]
It seems to work now. At least that part. I now see an issue which I remember having since I started using Soot, which is that with local.getName() I get the soot name but not the actual variable name.

I tried once compiling with the `-d` option but didn't work for me.

**Comment by patricklam** [comment]
Yeah you need to preserve original names (using the option) for that, and
they may not exist due to compilation.
